Improve performance by lazily unserializing and loading types from the schema

spawnia opened the pull request:
This PR brings a major performance boost to Lighthouse by utilizing [lazy type loading](http://webonyx.github.io/graphql-php/type-system/schema/#lazy-loading-of-types).

Real-world tests in our application have shown a total speed increase of about 4x. We have been using this branch in production for the last two weeks, it has been working great.

Here is the benchmark for how the new unserialization performs in relation to a naive,
full schema scan.

```
+------------------------------+-----------+
| subject                      | mean      |
+------------------------------+-----------+
| deep schema scan             | 149.697μs |
| DocumentAST before           | 46.169μs  |
| DocumentAST after            | 25.858μs  |
+------------------------------+-----------+
```

Resolves https://github.com/nuwave/lighthouse/issues/751
Resolves https://github.com/nuwave/lighthouse/issues/400
~~Probably also~~ Resolves https://github.com/nuwave/lighthouse/issues/636

**Breaking Changes**

Unfortunately, yes.

Even if we were to keep the API compatible, there might still be dependence on the fact that right now, the whole schema including all directives are eagerly evaluated.

The API for AST manipulation was completely reworked. Turns out that the functional paradigms we tried to apply (immutability & pure functions) were resulting in major performance degradations and were needlessly complicating the transformations. Thus, i converted all that to be explicitely mutating, the methods that deal with the AST signify that by using pass-by-reference `&parameters` and are returning `: void`.

## Changelog

### Added

- Add the `@namespace` directive as a replacement for the removed `@group` directive https://github.com/nuwave/lighthouse/pull/768

### Fixed

- Avoid growing the memory extensively when doing complex AST manipulation https://github.com/nuwave/lighthouse/pull/768

### Changed

- Bumped the requirement on `webonyx/graphql-php` to `^0.13.2` https://github.com/nuwave/lighthouse/pull/768
- Rename directive interfaces dealing with types from `Node*` to `Type*` https://github.com/nuwave/lighthouse/pull/768
- Change the signature of the AST manipulating directive interfaces: `TypeManipulator`, `FieldManipulator` and `ArgManipulator` https://github.com/nuwave/lighthouse/pull/768
- Change the API of the `DocumentAST` class to enable a more performant implementation https://github.com/nuwave/lighthouse/pull/768
- Enable the schema caching option `lighthouse.cache.enable` by default https://github.com/nuwave/lighthouse/pull/768 
- Lazily load types from the schema. Directives defined on parts of the schema that are not used within the current
  query are no longer run on every request https://github.com/nuwave/lighthouse/pull/768

### Removed

- Remove `@group` directive in favour of `@middleware` and `@namespace` https://github.com/nuwave/lighthouse/pull/768

chrissm79 commented:
Those benchmarks 😍 🎉

alberthaff commented:
Awesome. This looks very promising, @spawnia !

I have a bunch of custom directives that i have to rewrite in order to test this. I hope to get it done later today.

spawnia commented:
@alberthaff can you provide some examples of the changes you had to make to your directives?

alberthaff commented:
I had to among others, switch out `NodeManipulator` with `FieldManipulator` and change some of the code that generates/ adds types, fields and directives programatically, since `TypeRegistry::register` was removed.

I was still running the beta for version 3, so there has most likely been a lot of changes since then. :-)

olivernybroe commented on composer.json:
```diff
@@ -38,6 +38,7 @@
         "mockery/mockery": "^1.0",
         "orchestra/database": "3.5.*|3.6.*|3.7.*|3.8.x-dev",
         "orchestra/testbench": "3.5.*|3.6.*|3.7.*|3.8.*",
+        "phpbench/phpbench": "@dev",
```
Why do we require dev of phpbench?

spawnia replied:
Because that is what they recommend https://phpbench.readthedocs.io/en/latest/installing.html#composer-install

olivernybroe commented on docs/master/api-reference/directives.md:
````diff
@@ -804,6 +783,31 @@ type User @model {
+}
+```
+
+An `@namespace` directive defined on a field wins in case of a conflict.
````
```suggestion
A `@namespace` directive defined on a field directive wins in case of a conflict.
```

olivernybroe commented on src/Defer/Defer.php:
```diff
@@ -74,17 +74,17 @@ public function __construct(CanStreamResponse $stream, GraphQL $graphQL)
      * @return void
      */
-    public function handleManipulateAST(ManipulateAST $ManipulateAST): void
+    public function handleManipulateAST(ManipulateAST $manipulateAST): void
```
Maybe we should add in our contribution guide how variable naming is. Should this be named `manipulateAst` instead. For example in the `GraphQl.php` file we have a variable called `astBuilder`, if the naming should be the same as in here, it should be named `aSTBuilder`.

spawnia replied:
`aSTBuilder` just looks really weird to me. Variable naming is not that important, we just camelCase and call it a day :)

olivernybroe replied:
I agree, aSTBuilder looks really weird and is confusing. It looks like we are saying `a ST Builder`, not `AST Builder`. 
But great, sound perfect :)

olivernybroe commented on src/Pagination/PaginationManipulator.php:
```diff
@@ -47,15 +47,15 @@ public static function transformToPaginatedField(
-        DocumentAST $documentAST,
+        FieldDefinitionNode &$fieldDefinition,
+        ObjectTypeDefinitionNode &$parentType,
+        DocumentAST &$documentAST,
```
Why is is needed to pass them by reference?

alberthaff replied:
If we need to add custom types, they are added to `$documentAST->types`, which makes them available globally. I think that is why it is added with `&`.

spawnia replied:
Trying to shoehorn a non-mutable style into the AST manipulation has caused lots of additional complexity and issues such as #636 

It is not really needed to pass them by reference, since objects are mutable anyway, but i think it is more explicit and obvious that way.

olivernybroe replied:
I agree with keeping it mutable. Was just wondering if there were any reason for it, but I guess by reference is faster when we are dealing with an AST, as they are often pretty big. 
Not that the speed impact for that should matter :)

olivernybroe commented on benchmarks/ASTUnserializationBench.php:
```diff
@@ -0,0 +1,73 @@
+/**
+ * @BeforeMethods({"prepareSchema"})
+ */
+class ASTUnserializationBench
```
Could we add phpbench into automated tests? Just so we can see the results in a PR, not necessarily make rules for how it should perform.

spawnia replied:
I don't really want it in travis, as the benchmarks can get quite slow.

olivernybroe replied:
I have no idea about how to run it async, however travis should just use the return code from the command. So if we make it not return an error code, that might work.

Just thinking that it could be useful for us to see the results of the benchmark whenever a user makes a new PR.

spawnia replied:
Added it in as a job that is allowed to fail

spawnia replied:
https://travis-ci.org/nuwave/lighthouse/jobs/530344137

spawnia commented:
@alberthaff added TypeRegister->register back in. It's quite useful actually, just got around to using it myself.

nrobates commented:
This is great @spawnia. So far the benchmarks are showing an exact increase in performance equivalent to the decrease we experienced. Very excited for this.

spawnia commented:
@nrobates decrease in comparison too?

nrobates commented:
@spawnia I was just meaning that we noticed an increase in time by 2-4 times what it used to be which I referenced as decrease in performance. So your benchmarks are showing an increase in performance  and speed in offset.

olivernybroe commented on composer.json:
```diff
@@ -30,13 +30,12 @@
         "illuminate/validation": "5.5.*|5.6.*|5.7.*|5.8.*",
-        "opis/closure": "^3.0",
-        "webonyx/graphql-php": "^0.13"
+        "webonyx/graphql-php": "^0.13.2"
```
What feature are we requiring from 0.13.2 since you changed it? Isn't it better to have it as loose as possible, composer will automatically install the best version possible.

spawnia replied:
https://github.com/webonyx/graphql-php/pull/475

alberthaff commented:
@spawnia 
I apparently was not using the schema caching until now.

After the commit that enables caching by default, my endpoint returns an exception when the schema is loaded from cache.

```
(1/1) ErrorExceptionUndefined offset: 0
--
in NodeList.php line 57
at HandleExceptions->handleError(8, 'Undefined offset: 0', '/Users/alberthaff/project/vendor/webonyx/graphql-php/src/Language/AST/NodeList.php', 57, array('offset' => 0))in NodeList.php line 57
at NodeList->offsetGet(0)in NodeList.php line 118
at NodeList->getIterator()in TypeRegistry.php line 133
```

 It works fine for the first request after i clear the cache.

spawnia commented:
@alberthaff have you updated the dependency to `webonyx/graphql-php` as well? I know this error, it is fixed within `0.13.2`

alberthaff commented:
@spawnia That fixed it. Thanks!

spawnia commented:
@chrissm79 now that the docs update has landed, i have given this some final polish and fixed the merge. I think we are good to go here. Let's release this as an alpha right away.

I added a `3.x` branch if we require fixes to be backported while we are readying `master` for a full-blown `4.0`

chrissm79 commented:
Amazing work on this one @spawnia 👍Thank you!!!

mfn commented:
Wow, congratulations 🎉 Looks like it was almost an Herculean effort!

alberthaff commented:
Awesome work!

spawnia commented:
Released this as the first alpha for v4. https://github.com/nuwave/lighthouse/releases/tag/v4.0-alpha.1
